### DISCUSSION TITLE
Connecting to RPi Pico using Serial over TCP connection

### DISCUSSION THREAD

**Discussion by kc64** [BrianPugh/belay#171] [Q&A]
I have mounted a Pico to the USB port of a remote computer and then ran the tcp_serial_redirect tool that is part of the pySerial package. Using PuTTY, I can connect to it without issue using a raw socket, the remote machine address, and the port number I specified. Is there a connection method in belay to support this type of connection?

**Comment by BrianPugh** [comment]
Hey @kc64 !

I wasn't familiar with [`tcp_serial_redirect.py`](https://github.com/pyserial/pyserial/blob/master/examples/tcp_serial_redirect.py), but I was able to get it working without too much trouble (use the newly released version of Belay, v0.22.5).

Steps (running blink demo on a pi pico):
1. Modify the [blink example](https://github.com/BrianPugh/belay/blob/main/examples/01_blink_led/main.py) so that the device is created with a blank username. A blank username will cause telnet to connect without authentication:
   ```
   device = belay.Device(args.port, user="")
   ```

2. In the [pyserial/examples](https://github.com/pyserial/pyserial/tree/master/examples) directory, run:
   ```bash
   $ python tcp_serial_redirect.py /dev/ttyUSB0 115200  # replace device port with your device port
   --- TCP/IP to Serial redirect on /dev/cu.usbmodem12401  115200,8,N,1 ---
   --- type Ctrl-C / BREAK to quit
   Waiting for connection on 7777...
   ```
   
3. In a different terminal, at the root  belay directory, run the modified example:
   ```bash
   $ $ python examples/01_blink_led/main.py --port=127.0.0.1:7777
   Printing from device; turning LED to True.
   Printing from device; turning LED to False.
   Printing from device; turning LED to True.
   ```

And it should work! Note that the builtin `telnetlib` library was removed in the newly released python 3.13, so this telnet feature will not work on python 3.13. It seems like the solution is to just vendor the file into Belay. I'll do this if there's a need!

**Comment by kc64** [comment]
Hi @BrianPugh --

Again with the super fast, very helpful support. Many thanks for that. I've been using the new version with some success and have only had one problem that I can't explain presently--the Pico dropped into FS mode. I'll keep my eye on this one in case it recurs. Another observation is that if I connect to the Pico using Putty, there is no delay before I can start using the device. The REPL is immediately available. On the other hand, if I connect via a python script, there appears to be a 3-5 second delay before my `belay.Device.__call__` instances start working.

**Comment by BrianPugh** [comment]
Give v0.22.6 a try, it should be much much faster. I also vendored `telnetlib` in this release, so it's now python3.13 compatible.

**Reply by kc64** [reply]
It's working wonderfully. Many thanks!
